### PR TITLE
feat(fmt): support YAML

### LINKED ISSUE #12217: Deno fmt for yml files
Would that be something worth considering?

### PR DESCRIPTION
**Pull request by g-plane**
This PR will close #12217 which is a feature request that is asked nearly three years ago!

A question not related to this: I've also made a dprint plugin that can format CSS, Sass, SCSS and Less, and it's written in Rust. Can we consider integrating it to "deno fmt"?

### CODE REVIEW

File: cli/tools/fmt.rs [hunk 1]
```diff
@@ -687,6 +704,35 @@ fn get_resolved_json_config(
...
+    layout_options.use_tabs = use_tabs;
+  }
+  if let Some(line_width) = options.line_width {
+    layout_options.print_width = line_width as usize;
+  }
+  if let Some(indent_width) = options.indent_width {
+    layout_options.indent_width = indent_width as usize;
+  }
+
+  let mut language_options = LanguageOptions::default();
```
Comment (dsherret) [inline cli/tools/fmt.rs:723]: Can you initialize this like:

```rs
let language_options = LanguageOptions {
  // ...
};
````

...and list every property along with the defaults? (same with LayoutOptions) That way the defaults here never change here.

File: cli/tools/fmt.rs [hunk 2]
```diff
@@ -687,6 +704,41 @@ fn get_resolved_json_config(
...
+fn get_resolved_yaml_config(
+  options: &FmtOptionsConfig,
+) -> pretty_yaml::config::FormatOptions {
+  use pretty_yaml::config::*;
+
+  let layout_options = LayoutOptions {
+    print_width: options.line_width.unwrap_or(80) as usize,
+    use_tabs: options.use_tabs.unwrap_or_default(),
+    indent_width: options.indent_width.unwrap_or(2) as usize,
+    line_break: LineBreak::Lf,
```
Comment (petamoriken) [inline cli/tools/fmt.rs:716]:
```suggestion
    ..Default::default()
```

File: cli/tools/fmt.rs [hunk 3]
```diff
@@ -687,6 +704,41 @@ fn get_resolved_json_config(
...
+    },
+    trailing_comma: true,
+    format_comments: false,
+    indent_block_sequence_in_map: true,
+    brace_spacing: true,
+    bracket_spacing: false,
+    dash_spacing: DashSpacing::OneSpace,
+    trim_trailing_whitespaces: true,
+    trim_trailing_zero: false,
+    ignore_comment_directive: "deno-fmt-ignore".into(),
```
Comment (petamoriken) [inline cli/tools/fmt.rs:733]:
```suggestion
    ..Default::default()
```

File: cli/tools/fmt.rs (hunk 2 above)
Comment (dsherret) [reply cli/tools/fmt.rs:716]: See https://github.com/denoland/deno/pull/24717#discussion_r1691411918

Comment (petamoriken) [reply cli/tools/fmt.rs:716]: I had overlooked this 🙇

**Comment by bartlomieju** [comment]
> A question not related to this: I've also made a dprint plugin that can format CSS, Sass, SCSS and Less, and it's written in Rust. Can we consider integrating it to "deno fmt"?

Yes, please! This would be a fantastic addition. @dsherret also mentioned that you have an HTML formatter. We would be interested in integrating them all.

**Comment by bartlomieju** [comment]
I just checked and this PR increases binary size on mac by about 200Kb:
```
$ ls -la $(which deno)
-rwxr-xr-x@ 1 ib  staff  105250144 Jul 26 20:23 /Users/ib/.deno/bin/deno*
$ ls -la ./target/release/deno
-rwxr-xr-x@ 1 ib  staff  105459944 Jul 30 22:31 ./target/release/deno*
```

**Comment by ry** [comment]
20 bps is not nothing... I wonder how much the html or css formatters would add.

**Comment by g-plane** [comment]
We shouldn't just focus on increased size. Instead, we should consider "Is the increased size worth solving problems?".

The original issue was opened three years ago and received some upvotes. Meanwhile, the first version of Pretty YAML is released about one month ago, but its dprint plugin has been downloaded nearly 6000 times.
image
*source: https://plugins.dprint.dev/*

**Comment by bartlomieju** [comment]
@g-plane we're on the same page. Given the size increase of the YAML formatter we're feeling optimistic about CSS and HTML formatters. Would you be able to open PRs for them as well?

**Comment by g-plane** [comment]
Considering that it may have merge conflicts with this PR, when should I do? Should I wait for this PR? And, recently I'm working on some optimizations (not size-related) on CSS formatter, so it may happen several days later, at least.

**Comment by bartlomieju** [comment]
Good point! We're discussing if we'd like to add unstable flags for the formatters (eg. `--unstable-fmt-yaml`) for some time to collect the feedback from the users before fully committing. We will have a decision on that early next week and then we'll merge this PR. 

Let's wait with other PR until after this one is merged. 

We're quite interested in having YAML, CSS and HTML formatters shipped in Deno v1.46 so just let me know if we can help with any of the PRs.

**Review by dsherret — APPROVED** [review]
LGTM!

**Pull request merged**